src/platform/windows/mod.rs

```diff
@@ -0,0 +1,1431 @@
+            }
+            try!(write_buf(*self.handle, &oob_data));
+            if big_data_sender.is_some() {
+                try!(big_data_sender.unwrap().send_unfragmented(data));
+            }
+        }
+
+        Ok(())
+    }
+}
+
+pub struct OsIpcReceiverSet {
+    // the IOCP that we select on
+    iocp: WinHandle,
+
+    // the set of receivers
+    receivers: Vec<OsIpcReceiver>,
+}
+
+pub enum OsIpcSelectionResult {
+    DataReceived(i64, Vec<u8>, Vec<OsOpaqueIpcChannel>, Vec<OsIpcSharedMemory>),
+    ChannelClosed(i64),
+}
+
+impl OsIpcReceiverSet {
+    pub fn new() -> Result<OsIpcReceiverSet,WinError> {
+        unsafe {
+            let iocp = kernel32::CreateIoCompletionPort(INVALID_HANDLE_VALUE,
+                                                        ptr::null_mut(),
+                                                        0 as winapi::ULONG_PTR,
+                                                        0);
+            if iocp == ptr::null_mut() {
+                return Err(WinError::last("CreateIoCompletionPort"));
+            }
+
+            Ok(OsIpcReceiverSet {
+                iocp: WinHandle::new(iocp),
+                receivers: vec![],
+            })
+        }
+    }
+
+    pub fn add(&mut self, mut receiver: OsIpcReceiver) -> Result<i64,WinError> {
+        // use this to identify the receiver
+        let receiver_handle = *receiver.handle;
+
+        // XXX we'll need a mutex here... at least while we loop through
+        // receivers to find a matching handle when we get a IOCP
+        try!(receiver.add_to_iocp(*self.iocp));
+        self.receivers.push(receiver);
+
+        dd!("[# {:?}] ReceiverSet add {:?}", *self.iocp, receiver_handle);
+
+        Ok(receiver_handle as i64)
+    }
+
+    pub fn select(&mut self) -> Result<Vec<OsIpcSelectionResult>,WinError> {
+        assert!(self.receivers.len() > 0, "selecting with no objects?");
+        dd!("[# {:?}] select() with {} receivers", *self.iocp, self.receivers.len());
+
+        unsafe {
+            // the ultimate results
+            let mut selection_results = Vec::new();
+
+            // make a local var for this, because in the future
+            // we'll take it from a mutex
+            let ref mut receivers = &mut self.receivers;
+
+            // Make a quick first-run check for any closed receivers.
+            // This will only happen if we have a receiver that
+            // gets added to the Set after it was closed (the
+            // router_drops_callbacks_on_cloned_sender_shutdown test
+            // causes this.)
+            receivers.retain(|ref r| {
+                if r.is_closed() {
+                    selection_results.push(OsIpcSelectionResult::ChannelClosed(*r.handle as i64));
+                    false
+                } else {
+                    true
+                }
+            });
+
+            // if we had prematurely closed elements, just process them first
+            if selection_results.len() > 0 {
```

Comment:
(BTW, there is `.is_empty()` if you'd like to use it for stuff like this; up to you)